frontend/src/metabase/query_builder/selectors.js

```diff
@@ -637,12 +635,15 @@ export const getShouldShowUnsavedChangesWarning = createSelector(
 
     if (isNative) {
       const isNewQuestion = !originalQuestion;
+      const rawQuery = Lib.rawNativeQuery(question.query());
 
       if (isNewQuestion) {
-        return !question.legacyQuery().isEmpty();
+        return rawQuery.length > 0;
       }
 
-      return isSavedQuestionChanged;
+      const rawOriginalQuery = Lib.rawNativeQuery(originalQuestion.query());
+      const hasQueryChanged = rawQuery !== rawOriginalQuery;
+      return hasQueryChanged;
```

Comment:
This is incorrect due to references getting expanded in some cases.
Luckily e2e tests caught that.

![image](https://github.com/metabase/metabase/assets/6830683/898e5412-52e1-42fa-a69f-046d091f5ec8)

Reply:
Fixed by 16704821e6a912775cf4ac798aa89bf0b894440b